src/ProjectSystem/Impl/IO/MsBuildFileSystemWatcher.cs

```diff
@@ -215,6 +215,9 @@ public sealed partial class MsBuildFileSystemWatcher : IDisposable {
 
             relativePath = PathHelper.MakeRelative(rootDirectory, fullPath);
             try {
+                if (!fileSystem.FileExists(fullPath)) {
```

Comment:
Not entirely. FileExists check prevents returning false in IsFileAllowed. Returning false may actually explain when sometimes file is added, but does not appear in project explorer. Race occurs when check happens between file is deleted and then re-created. I guess the event queue is not compact i.e. create-delete-create is not collapsed into just created.